lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1874,51 +2131,163 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
       // Until Swift 6, only emit a warning when we get this with an
       // explicit capture, since we used to not diagnose this at all.
       auto shouldOnlyWarn = [&](Expr *selfRef) {
+        auto DRE = dyn_cast_or_null<DeclRefExpr>(selfRef);
```

Comment:
This local function is over a hundred lines long, and it's inside a method that's even longer. I feel like this 'implicit self' diagnostic does not carry its weight, because it introduces a lot of new bookkeeping in name lookup, and the diagnostic logic here is also quite complex. Is there any way we can simplify the rule? Is the diagnostic worth having at all?

Reply:
For example, here's an idea: instead of "shouldOnlyWarn", why not make it always a warning? Then you can delete some code...

Reply:
In terms of whether or not the diagnostic should exist at all: it implements functionality defined in [SE-0269](https://github.com/apple/swift-evolution/blob/main/proposals/0269-implicit-self-explicit-capture.md) and [SE-0365](https://github.com/apple/swift-evolution/blob/main/proposals/0365-implicit-self-weak-capture.md), which meaningfully improve the ergonomics of working with closures, and have been accepted for some time (but with bugs in the existing implementations).

Reply:
> If we wanted to simplify the code we could deprioritize some of the edge cases and accept these downgraded warnings in more cases in Swift 5 mode.

My overall concern is that the existence of these edge cases suggests something isn't being encoded or analyzed properly and we're probably missing a simpler model that would produce the same diagnostics. A reasonable way to start is by splitting up some of these longer functions so that it's easier to tell what's going on. I also find that nested local functions can become unreadable when they're longer than, say, 20 lines of code each. It might even be worth moving some stuff out of MiscDiagnostics.cpp into a new file.

I'm fully aware of the fact that MiscDiagnostics.cpp is already a huge tangled mess -- after all, we're diagnosing a bunch of stuff that just doesn't fit anywhere else -- but that's all the more reason to tidy things up. I want to say I definitely appreciate your efforts, and it would be great to chip away at the complexity a little bit more.

You can do any of this in a follow-up PR since you've been iterating on this one for a while already.

Also FWIW, I recently added this FIXME in `diagnoseImplicitSelfUseInClosure()`:

```
      // FIXME: This is happening too early, because closure->getType() isn't set.
      if (auto *closure = dyn_cast<AbstractClosureExpr>(DC))
        if (closure->getType())
          if (DiagnoseWalker::isClosureRequiringSelfQualification(closure, ctx))
            ACE = const_cast<AbstractClosureExpr *>(closure);
```

The getType() check was hidden behind a utility function before, and it would silently skip the isClosureRequiringSelfQualification() check if the closure hadn't been type checked yet. This suggests the visitor is looking at too many things, or looking at things in the wrong order. If you change the check into an assertion you'll see the problem.

Reply:
> without unintentionally downgrading any existing errors to warnings.

This is a noble goal but I wouldn't prioritize it over all else. It's always safer to just downgrade to a warning.

Reply:
Thanks -- for this PR I'll at least reorganize the code to be easier to follow, and do a pass over each of the checks in `shouldOnlyWarn` to see if any are good candidates to be removed for simplicity.

Reply:
Thanks for the advice @slavapestov! I refactored and simplified the `shouldOnlyWarn` logic in https://github.com/apple/swift/pull/70575/commits/d17bc35a93a6b9073162ceea80704324e8beeec4. The code is much better organized now, so should be much easier to follow.

I reviewed each of the individual checks, and was able to remove a few. The checks that remain are all targeted and well motivated (e.g. required to preserve source compatibility).